### PR TITLE
fix: seeds in `TorchAlgo`, Camelyon reproducibility issues and performance controls

### PR DESCRIPTION
**Pull request by thbcmlowk**
## Related issue

Closes FL-1095

Closes FL-1123
 
## Summary

* Improves handling of benchmark results and summaries
* Adds control on benchmark results:
  * Checks `substrafl` and `torch` performances are equals
  * Checks `substrafl` and `torch` performances results match the expected results
* Fix multiple seed issues
* Fix model reinitialization issues 
* Minor FL improvements

### Added

`benchmark_metrics`

  * `BenchmarkResults` responsible for structuring benchmak results
  * `PerformanceError(Exception)` to raise dedicated benchmark-related performance errors
  * `get_metrics_from_substra_performances` to cast `substra.sdk.models.Performances` into `BenchmarkMetrics`
  * Assertion methods on performances
    * Ensures number of sets on which metrics are computed is the same
    * Ensures computed metrics are the same on each client
    * Ensures results of each metrics is the same on each experiement
    * Ensures results match the expected ones according to the run mode (subprocess or remote)

### Changed

  * `substrafl.algorithms.pytorch.torch_base_algo`:
    * Seed `random.seed`
    * Seed `np.random.seed`
  
  * `benchmarks`:  
    * Integrates `benchmark_results`        
    * Systematically loads the benchmark summary into the `results` folder
    * Fix run artifacts cleaning issues
  
  * `utils.read_results` → `utils.load_benchmark_summary` which is now responsible for loading the reformatted (see appendix) json summary according to a number of limit experiment to keep
  
  * weldon:
    * Fix seed usage and model reset 
    * Update loss function `torch.nn.BCEWithLogitsLoss` → `torch.nn.BCELoss`    
    * Update predictions by adding sigmoid and handle reshaping at lower level

## Proof of Concept

See `PerformanceError` raised before fixing performance issues: https://github.com/Substra/substrafl/actions/runs/5596758984/jobs/10234206533?pr=151

Tested locally with N=20 repetitions to ensure reproducibility.

## Appendix

Snippet of the reformatted `results/results.json` summary file:
```json
    {
        "2023-08-07T09:43:55": {
            "asset_keys": "keys.json",
            "batch_size": 4,
            "credentials": "remote.yaml",
            "learning_rate": 0.01,
            "mode": "subprocess",
            "n_centers": 2,
            "n_local_steps": 1,
            "n_rounds": 2,
            "nb_test_data_samples": 2,
            "nb_train_data_samples": 2,
            "num_workers": 0,
            "results": {
                "substrafl": {
                    "exec_time": 17.73998,
                    "n_clients": 2,
                    "performances": {
                        "Accuracy": 0.5,
                        "ROC AUC": 1.0
                    }
                },
                "torch": {
                    "exec_time": 1.7152729034423828,
                    "n_clients": 2,
                    "performances": {
                        "Accuracy": 0.5,
                        "ROC AUC": 1.0
                    }
                }
            },
            "seed": 42,
            "substra_version": "0.45.0",
            "substrafl_version": "0.38.0",
            "substratools_version": "0.20.0"
        }
    },
```

## Please check if the PR fulfills these requirements

- [x] If the feature has an impact on the user experience, the [changelog](https://github.com/substra/substrafl/blob/main/CHANGELOG.md) has been updated
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] The commit message follows the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) specification

### CODE REVIEW

**Comment by thbcmlowk** [comment]
/e2e --tests sdk,substrafl,mnist,camelyon --orchestrator-mode standalone

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by ThibaultFy** [comment]
Is this ready to review ? If yes, we usually open the PR to emphasise this :)

**Comment by thbcmlowk** [comment]
> Is this ready to review ? If yes, we usually open the PR to emphasise this :)

Nope, sorry! That's why it has been switched to draft again. 
Still having reproducibility issues (even with hard seeding as suggested in https://github.com/Substra/substrafl/pull/152)

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by thbcmlowk** [comment]
/e2e --tests=sdk,substrafl --benchmarks=camelyon --orchestrator-mode=standalone

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: benchmark/camelyon/benchmark_metrics.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    """
+    execution_time: datetime.timedelta = performances.compute_plan_end_date[0] - performances.compute_plan_start_date[0]
+    execution_time: float = execution_time.total_seconds()
+
+    workers, metrics, values = performances.worker, performances.identifier, performances.performance
+    performances = {}
+
+    for org, metric, value in list(zip(workers, metrics, values)):
+        if org not in performances:
+            performances[org] = {}
```
Comment (SdgJlbl) [inline benchmark/camelyon/benchmark_metrics.py:127]: Wondering if it would be more readable to use `defaultdict(dict)` for `performances` instead 🤔 
(But I might tend to overuse defaultdict, so no strong opinion here)

**Review by SdgJlbl — COMMENTED** [review]
Thanks a lot for the PR
A few minor comments, the only thing I'm really wondering about is that the metrics seem to be different depending on the mode used (unless I'm missing something).

File: benchmark/camelyon/common/utils.py [hunk 2]
```diff
@@ -91,19 +92,19 @@ def parse_params() -> dict:
...
-        results = {}
-
-    return results
+    """
+    n_experiment_limit = -1 * max(n_experiment_limit, 0)
+    file.parent.mkdir(exist_ok=True)
+    experiment_summary = {datetime.datetime.utcnow().strftime("%Y-%m-%dT%H:%M:%S"): experiment_summary}
+    experiment_summary = (json.loads(file.read_text()) if file.exists() else []) + [experiment_summary]
+    file.write_text(json.dumps(experiment_summary[n_experiment_limit:], sort_keys=True, indent=4))
+    return
```
[the remark this answers is not in the document]
Comment (thbcmlowk) [reply benchmark/camelyon/common/utils.py:110]: Sorry, old habits of having explicit `return` statements even when nothing is returned. Will try to remember the convention here. Thank you!

File: benchmark/camelyon/benchmarks.py [hunk 3]
```diff
@@ -52,46 +50,33 @@ def fed_avg(params: dict, train_folder: Path, test_folder: Path):
         "num_workers",
         "nb_train_data_samples",
         "nb_test_data_samples",
+        "seed",
     ]
 
-    cl_start = time.time()
-    cl_perf = substrafl_fed_avg(
+    cl_metrics = substrafl_fed_avg(
```
[the remark this answers is not in the document]
Comment (thbcmlowk) [reply benchmark/camelyon/benchmarks.py:56]: I did not get that 😄 definitely agree to rename.

File: benchmark/camelyon/benchmark_metrics.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    Returns:
+        None
+
+    Raises:
+        PerformanceError if results are not the one expected.
+    """
+    cl_metrics.assert_performance_equals(other=sa_metrics)
+
+    if mode == "subprocess":
+        expected_performances = {"Accuracy": 0.5, "ROC AUC": 1.0}
```
[the remark this answers is not in the document]
Comment (thbcmlowk) [reply benchmark/camelyon/benchmark_metrics.py:91]: My mistake, I think I went too fast on this part.

The idea is that when using the Makefile entrypoints, one will run either:
* The `"subprocess"` mode with:
		--nb-train-data-samples 2 \
		--nb-test-data-samples 2 \
		--batch-size 4 \
		--n-local-steps 1 \
		--n-rounds 2
* The `"remote"` mode with:
		--nb-train-data-samples 5 \
		--nb-test-data-samples 2 \
		--batch-size 4 \
		--n-local-steps 10 \
		--n-rounds 7

The expected results does not depend on the mode but on the underlying parameters (number of samples, steps, rounds...). Definitely needs an update, but not sure on how to infer the expected results robustly...

[the remark this answers is not in the document]
Comment (SdgJlbl) [reply benchmark/camelyon/benchmark_metrics.py]: That's very clear, thanks for the explanation.

**Review by SdgJlbl — APPROVED** [review]
Thanks

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**
